sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -95,13 +95,11 @@ private[sql] object Dataset {
   def ofRows(sparkSession: SparkSession, logicalPlan: LogicalPlan): DataFrame =
     sparkSession.withActive {
       val qe = sparkSession.sessionState.executePlan(logicalPlan)
-      val encoder = if (qe.isLazyAnalysis) {
-        RowEncoder.encoderFor(new StructType())
-      } else {
+      if (!qe.isLazyAnalysis) {
         qe.assertAnalyzed()
-        RowEncoder.encoderFor(qe.analyzed.schema)
       }
-      new Dataset[Row](qe, encoder)
+      val encoder = () => RowEncoder.encoderFor(qe.analyzed.schema)
+      new Dataset[Row](qe, encoder())
```

Comment:
Does it mean we define a function and then invoke the function immediately?

Reply:
No, the signature of the constructor changed to `lazyEncoder: => Encoder[T]`, so it should be invoked later.
I changed it to make it clearer and to fix the test failures at 584c650.